### DISCUSSION TITLE
Resource temporarily unavailable errors

### DISCUSSION THREAD

**Discussion by jault3** [Kong/kong#10662] [Help] [answered]
👋  I recently upgraded Kong OSS Gateway from 2.8.3 to 3.0.2 and Kubernetes Ingress Controller 2.7.0. Ever since then, I've been encountering an error that I have been unable to resolve. The proxy container frequently logs this message
```
ingress-kong-66b654c779-msc2s proxy 2023/04/12 20:24:55 [crit] 1127#0: *5749284 connect() failed (11: Resource temporarily unavailable), context: ngx.timer
```
It happens every couple minutes, and when it happens, it logs that same message over 6,000 times in a 30 second time frame. I'm also seeing semi-regular reports from other team members of Kong returning the `{"message":"failure to get a peer from the ring-balancer"}` error message across many different services.

I'm not sure what this resource unavailable error means (couldn't find any other examples of anyone having a similar issue), or what could be causing it, and I'm hoping this is familiar to someone who can point me in the right direction so I can gather more info and open an issue if needed!

So far I have tried increasing the `mem_cache_size` to `500m` (from 256m), upping the `net.core.somaxconn` setting on the VM to match the default backlog value of 16384 (VM defaults to 1024), tripling the number of replicas (from 2 to 6), dropping the database entirely and bootstrapping it from scratch, and raising both the CPU and memory requests & limits. Unfortunately nothing seems to have helped much. I am also running the exact same configuration in 4 other Kubernetes clusters that don't seem to have this issue (although they are all much smaller clusters).

Here's some potentially relevant details about my deployment:
* Runs in a kubernetes cluster & uses a postgres database
* This cluster contains 1048 ingresses, 1922 services, and 1310 pods
* Nearly all of the services in the cluster set the `ingress.kubernetes.io/service-upstream: "true"` annotation

I've also seen these errors fairly frequently, that seem to be related (just based on the fact that I'd never seen them before the upgrade):

```
ingress-kong-66b654c779-s44hs proxy 2023/04/12 20:24:59 [error] 1127#0: *756 [lua] targets.lua:183: on_target_event(): target create: balancer not found for c60d6c6a-0bae-4ea8-b857-81bebcc115b4, context: ngx.timer
...
ingress-kong-66b654c779-s44hs proxy 2023/04/12 20:36:09 [warn] 1128#0: *6293495 stream [lua] targets.lua:492: queryDns(): querying dns for <appName>.<ns>.svc failed: dns server error: 3 name error. Tried ["(short)<appName>.<ns>.svc:(na) - cache-miss","<appName>.<ns>.svc.kong.svc.cluster.local:33 - cache-miss/scheduled/try 1 error: failed to create a resolver: failed to set peer name: resource temporarily unavailable...
```
The `<appName>.<ns>` piece in that log message ^ changes quite a bit across log messages and doesn't seem to impact any one service more than others.

Thanks in advance for any help & pointers.

**Comment by ghost** [comment]
> It happens every couple minutes, and when it happens, it logs that same message over 6,000 times in a 30 second time frame. I'm also seeing semi-regular reports from other team members of Kong returning the `{"message":"failure to get a peer from the ring-balancer"}` error message across many different services.

https://github.com/Kong/kong/issues/3518
https://github.com/Kong/kong/issues/4778

Usually this means there are no healthy targets available upstream. I feel like this is related to DNS query errors below.


> ```
> ingress-kong-66b654c779-s44hs proxy 2023/04/12 20:36:09 [warn] 1128#0: *6293495 stream [lua] targets.lua:492: queryDns(): querying dns for <appName>.<ns>.svc failed: dns server error: 3 name error. Tried ["(short)<appName>.<ns>.svc:(na) - cache-miss","<appName>.<ns>.svc.kong.svc.cluster.local:33 - cache-miss/scheduled/try 1 error: failed to create a resolver: failed to set peer name: resource temporarily unavailable...
> ```

`<appName>.<ns>.svc` It looks like you are using k8s short domain name, can you switch to FQDN?

**Reply by jault3** [reply]
Thanks for the links! That was my hunch as well (for the ring-balancer errors). From what I can tell though, the application upstream is healthy. I checked the kubernetes events, described the pod, and it doesn't show anything out of the ordinary, such as failed healthchecks or restarts.

I also assumed Kong would defer to Kubernetes for load balancing based on the `ingress.kubernetes.io/service-upstream: "true"` annotation on the target service (but admittedly im not entirely sure how that behavior works and if Kong does any sort of load balancer still even with that annotation).

It's also definitely possible the ring-balancer errors are unrelated to the `Resource temporarily unavailable` errors, and some applications are unhealthy upstream (I only spot checked one of them, will look at more). But those resource temporarily unavailable errors are my main concern since they seem to happen almost constantly. I mostly included the ring-balancer errors in case they were in some way related. I will say that I can downgrade Kong back to 2.8.3 and all these errors do stop without any other changes in the cluster, which is very strange.

**Reply by jault3** [reply]
I forgot to mention, you are correct. We're using k8s short names. I'll try switching a couple to FQDN and report back

**Reply by ghost** [reply]
> I also assumed Kong would defer to Kubernetes for load balancing based on the `ingress.kubernetes.io/service-upstream: "true"` annotation on the target service (but admittedly im not entirely sure how that behavior works and if Kong does any sort of load balancer still even with that annotation).

If I remember correctly, as long as the upstream is set to a domain name, Kong will always resolve the domain name to IP address and then use the IP address to connect to the upstream. This is because `lua-resty-balancer` in OpenResty only accepts IP addresses. Therefore, if there is a problem with the domain name resolution of the upstream and no IP address is resolved, Kong will consider that there is no suitable upstream target available.

**Reply by jault3** [reply]
I looked into this a bit. I don't believe I can switch to using anything other than short service names. Since I use the Kong ingress controller, Kong is just reading ingress resources and configuring everything based on those. Ingresses only allow specifying the name of the service to forward traffic to in the same namespace as the ingress resource. So if I wanted to switch to different names, I am guessing that would have to be a function/configuration of the KIC. I didn't see anything explicitly in the docs about that, but maybe I missed it?

> If I remember correctly, as long as the upstream is set to a domain name, Kong will always resolve the domain name to IP address and then use the IP address to connect to the upstream

Good point, that makes sense. With `ingress.kubernetes.io/service-upstream: "true"`, I believe it should be resolving to a single cluster IP of the service resource, rather than load balancing across a set of pod IPs.

But in any case, I was able to get the DNS errors to stop entirely. I switched from 4 nginx workers down to 2, am now configuring the `net.core.somaxconn` sysctl value to be `16384`, and set both `KONG_NGINX_MAIN_WORKER_RLIMIT_NOFILE` and `KONG_NGINX_EVENTS_WORKER_CONNECTIONS` to be `32768`. The only errors I really see now are the `connect() failed (11: Resource temporarily unavailable)` ones. They are far less frequent than before though. It logs about 2,000 of them every half hour instead of 6,000 every 30 seconds.

**Reply by ghost** [reply]
Let's wait and see what other people familiar with k8s have to say about it.

**Reply by jault3** [reply]
I've been digging in to this more and think I found something else. I hopped on one of my kong pods and noticed there were thousands of open UDP connections (16,889 to be exact), while there were only 907 open TCP connections (ran `netstat` with the `-u` or `-t` flag). It looks like all the UDP connections are open to the kube-dns pods to perform DNS resolution. Here's an example of what all the logs look like:

```
udp        0      0 ingress-kong-74d8f6b977-mdgkp:30995 kube-dns.kube-system.svc.cluster.local:domain ESTABLISHED
```

So it seems like maybe theres a change in UDP connections? Or specific to DNS resolution within Kong? I checked my other clusters that are not having these issues and the UDP connection count is also very high (10,000) compared to TCP (655). So I think that the smaller scale of those other clusters is what's keeping things in check. However, I also checked my production clusters that are still running Kong 2.8.3 and those have only 3-5 open UDP connections to kube-dns.

I reviewed the release notes and nothing really stuck out to me, but is it possible that Kong is now opening and maintaining a persistent connection to the DNS server for every single upstream service? I was going to try quadrupling the `dns_cache_size` parameter (40,000) and expanding the `net.ipv4.ip_local_port_range` parameter described [here](https://www.nginx.com/blog/tuning-nginx/).

**Reply by ghost** [reply]
Looks like you're having a similar problem? #9959

It appears that Kong has executed too many DNS queries (UDP) and has not released the connections in a timely manner.

**Reply by ghost** [reply, marked as answer]
and maybe releated to: #10691

**Reply by jault3** [reply]
https://github.com/Kong/kong/pull/10691 - that looks like exactly what we're running in to. I'll keep an eye on that PR (and the upstream openresty one), and attempt another upgrade once that's merged in. Thanks for your help with this!
